userspace/ksud/src/uid_scanner.rs

```diff
@@ -1,88 +1,433 @@
+    fs::set_permissions(PID_FILE_PATH, perms)
+        .with_context(|| "failed to set PID file permissions")?;
+
+    info!("uid_scanner: PID file written: {} (pid={})", PID_FILE_PATH, pid);
+    Ok(())
+}
+
+#[cfg(target_os = "android")]
+fn setup_signal_handler() -> Result<()> {
     unsafe {
-        cmd.pre_exec(|| {
-            libc::nice(-20);
-            libc::setsid();
-            Ok(())
-        });
+        if libc::signal(libc::SIGUSR1, scan_signal_handler as usize)
+            == libc::SIG_ERR
+        {
+            anyhow::bail!("failed to set SIGUSR1 handler");
+        }
+        info!("uid_scanner: SIGUSR1 signal handler installed");
+    }
+    Ok(())
+}
+
+fn perform_scan_update(cfg: &ScannerConfig) {
+    match perform_uid_scan(cfg) {
+        Ok(_) => info!("uid_scanner: scan completed successfully"),
+        Err(e) => error!("uid_scanner: scan failed: {e}"),
+    }
+}
+
+pub fn run_daemon() -> Result<()> {
+    let dir = Path::new(USER_UID_BASE_DIR);
+    ensure_directory_exists(dir)?;
+
+    // Relax directory perms for kernel and other tools
+    let mut perms = fs::metadata(dir)?.permissions();
+    perms.set_mode(0o777);
+    fs::set_permissions(dir, perms)?;
+
+    // Generate random process name and set it
+    #[cfg(target_os = "android")]
+    {
+        let random_name = generate_random_process_name();
+        set_process_name(&random_name)?;
+        info!("uid_scanner: process name set to: {}", random_name);
     }
 
-    match cmd.spawn() {
-        Ok(child) => {
-            info!("uid scanner daemon started with pid: {}", child.id());
-            std::mem::drop(child);
+    // Write PID file so kernel can find us
+    write_pid_file()?;
+
+    #[cfg(target_os = "android")]
+    setup_signal_handler()?;
+
+    info!("uid_scanner: daemon starting");
+
+    let mut cfg = load_config().unwrap_or_default();
+
+    let mut kernel_enabled = is_kernel_enabled();
+
+    if kernel_enabled {
+        if cfg.auto_scan {
+            perform_scan_update(&cfg);
+        } else {
+            info!("uid_scanner: auto_scan disabled, waiting for manual or kernel requests");
         }
-        Err(e) => warn!("failed to start uid scanner daemon: {e}"),
+    } else {
+        info!("uid_scanner: kernel disabled, initial scan skipped");
     }
 
+    loop {
+        // Reload config & kernel flag periodically to pick up runtime changes
+        if let Ok(new_cfg) = load_config() {
+            cfg = new_cfg;
+        }
+        kernel_enabled = is_kernel_enabled();
+
+        // Check for kernel-initiated scan request via signal
+        let kernel_request = KERNEL_SCAN_REQUEST.swap(false, Ordering::Relaxed);
+        if kernel_request {
+            info!("uid_scanner: kernel scan request received via signal");
+        }
+
+        if kernel_enabled && (cfg.auto_scan || kernel_request) {
+            perform_scan_update(&cfg);
+        }
+
+        thread::sleep(Duration::from_secs(cfg.scan_interval_secs));
+    }
+}
```

Comment:
```suggestion
    loop {
        thread::park();
    }
}
```
you can just change there to infinity sleep, and eveything write to signal process, because signal will break your sleep status and force jump code to signal process and back to there